primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+    Output: Sky coordinate of the phase centre
+    '''
+    images = read_fits(path)
+    # Get pointing centre of observation
+    phase_centre_ra = images.header['OBSRA']
+    phase_centre_dec = images.header['OBSDEC']
+    # Convert to astropy.coordinates.SkyCoord object
+    phase_center = coordinates.SkyCoord(phase_centre_ra, phase_centre_dec, unit=(u.deg, u.deg))
+    # Get positions of other sources
+    source_pos = wcs.WCS(images.header)
+    return phase_center, source_pos
+
+
+def radial_offset(phase_center, source_pos, row, col):
+    '''
+    This function gets the position in ra and dec of a pixel and determine its
+    radial offset from phase centre.
+    Input: phase center and source positions
+    output: A separation vector between phase centre and source positions in radians.
+    '''
+    p2w = source_pos.pixel_to_world(row, col, 0, 0)[0]
+    separation_rad = p2w.separation(phase_center).rad
+    return separation_rad
+
+
+def central_freq(path):
+    '''
+    Function to get central frequency of each frequency plane.
+    Input : image header
+    Output : List of central frequency in MHz of each frequency plane.
+    '''
+    images = read_fits(path)
+    c_freq_plane = []
+    for i in range(1, images.header['NSPEC']+1):
+        c_freq_plane.append(images.header['FREQ{0:04}'.format(i)])
+    return np.array(c_freq_plane)
+
+
+def beam_pattern(separation_rad, path, c_freq=None):
+    '''
+    Making beam pattern using the DEEP2 paper
+    Input : Radial seperation array and path to the fits file
+    Output : Beam pattern
+    '''
+    if c_freq:
+        rho = separation_rad
+        c = c_freq*1e6
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        nu = c/1.e9
+        theta_b = vh_beam_mean*(1./nu)
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    else:
+        rho = separation_rad
+        c = central_freq(path)
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        for nu in c:
+            nu = nu/1.e9
+            theta_b = vh_beam_mean*(1./nu)
+            ratio = rho/theta_b
+            num = np.cos(1.189*np.pi*ratio)
+            dem = 1-4*(1.189*ratio)**2
+            a_b = (num/dem)**2
+            flux_density.append(a_b)
+    return flux_density
```

Comment:
There is an enormous amount of code repetition here, which it seems to me is just to check whether a single frequency or a list of frequencies is required.

I would separate out two functions:
Function 1:
Celled something like `cosine_power_pattern` that does the job of computing the power pattern for a given frequency and vectorised over an input list of separations.
Function 2:
Called something like `beam_pattern` that takes a fits file path, computes the separations from the `radial_offset` function and the central freuqencies from the `central_freq` function and returns a list of beams in each frequency slice in the fits header that was given.

That ought to simplify the flow of the `main()` function as well - since you don;t have to call the `radial_offset` there.

Reply:
It makes sense. Done.